### DISCUSSION TITLE
In-place assignment for awkward  array?

### DISCUSSION THREAD

**Discussion by Duchstf** [scikit-hep/awkward#609] [Q&A] [answered]
Hello,

so suppose that I want to do this with an awkward array:

```
a[a>4000] = 4000
```

How would I do this? Is there any current support for this operation?

Thanks!!

**Comment by jpivarski** [comment, marked as answer]
In-place assignment isn't supported as a design choice. There are two corners of "parameter space" we could have chosen:

   * allow in-place assignment and defensively copy arrays in complex operations, so that assignments don't lead to surprising long-distance consequences;
   * forbid in-place assignment and view, rather than copy, for most operations.

I chose the latter (after initial experience with an early version that _did_ allow in-place assignment). Defensive copies would be prohibitive for large data structures, such as records with many fields (which are common). The choice to make everything immutable was made _for performance_ (both speed and memory), which might sound surprising, considering that in-place mutations are used in NumPy for performance reasons (both speed and memory).

NumPy has this problem, too, but not as acutely. Some operations cause views, some cause copies:

```python
>>> import numpy as np
>>> original = np.array([1.1, 2.2, 3.3, 4.4, 5.5, 6.6, 7.7, 8.8, 9.9])
>>> sliced = original[::2]
>>> advanced = original[[0, 2, 4, 6, 8]]
>>> sliced
array([1.1, 3.3, 5.5, 7.7, 9.9])
>>> advanced
array([1.1, 3.3, 5.5, 7.7, 9.9])
>>> original[6] = 999
>>> sliced     # this is a view; it has changed
array([  1.1,   3.3,   5.5, 999. ,   9.9])
>>> advanced   # this is a copy; it has NOT changed
array([1.1, 3.3, 5.5, 7.7, 9.9])
```

(I think PyTorch has taken a positive step by making all the view operations have a different naming convention from the copy ones.)

This is a bit of an issue in NumPy because you have to be careful to check for view-vs-copy. It's endemic in Pandas (search for "SettingwithCopyWarning"). But it's harder in Awkward Array because rather than having one buffer that might be a view or might be a copy, almost all operations give you a tree with buffers attached to all the nodes of that tree in which some nodes are views and other nodes are new buffers. Which are views and which are new buffers is subject to change. [This PR](https://github.com/scikit-hep/awkward-1.0/pull/261), for example.

Therefore, the Awkward Array library itself does not include any operations that change the values of these buffers in place. You can do this kind of assignment in place:

```python
>>> import awkward as ak
>>> original = ak.Array([{"x": 1}, {"x": 2}, {"x": 3}])
>>> original["y"] = 10
>>> original
<Array [{x: 1, y: 10}, ... {x: 3, y: 10}] type='3 * {"x": int64, "y": int64}'>
```

but that's actually not changing any _buffers_ in place: it's creating a new tree structure with the new buffer (`y = 10`) added. If you have any other references to parts of `original` elsewhere, they are unharmed.

The other exception is that while Awkward Array doesn't define any in-place operations, nothing is stopping you from casting an Awkward Array (or part of one) as a NumPy array and changing it in place. This can potentially have long-range consequences, so if you do this, you'll have to be aware of its history. For example, it's fine to change in place an array that you have just created—you know exactly where it's been.

In the online documentation, [Mutability of Awkward Arrays from NumPy](https://awkward-array.org/how-to-convert-numpy.html#mutability-of-awkward-arrays-from-numpy) and [Mutability of Awkward Arrays converted to NumPy](https://awkward-array.org/how-to-convert-numpy.html#mutability-of-awkward-arrays-converted-to-numpy) discusses how to do this and what the issues are. Note that you can also cast Awkward Arrays as NumPy arrays in Numba and assign to them (PR #550).

**Comment by jpivarski** [comment]
I also should have given you a direct answer to your question:

```python
>>> a = ak.Array([1000, 2000, 3000, 4000, 5000, 6000])
>>> a
<Array [1000, 2000, 3000, 4000, 5000, 6000] type='6 * int64'>
>>> np.asarray(a)[a > 4000] = 4000
>>> a
<Array [1000, 2000, 3000, 4000, 4000, 4000] type='6 * int64'>
```

**Reply by jpivarski** [reply]
Here's an example of how it gets confusing quickly.

This `a` is a [ListOffsetArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListOffsetArray.html), for which the [ak.flatten](https://awkward-array.readthedocs.io/en/latest/_auto/ak.flatten.html) operation passes through a view:

```python
>>> a = ak.Array([[1, 2, 3], [], [4, 5]])
>>> np.asarray(ak.flatten(a))[2:] = 10
>>> a
<Array [[1, 2, 10], [], [10, 10]] type='3 * var * int64'>
```

This `b` is a [ListArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListArray.html), for which the [ak.flatten](https://awkward-array.readthedocs.io/en/latest/_auto/ak.flatten.html) returns a new buffer. When you change that new buffer in place, the original array is not changed in place.

```python
>>> b = ak.Array([[1, 2, 3], [], [4, 5]])[[0, 1, 2]]
>>> np.asarray(ak.flatten(b))[2:] = 10
>>> b
<Array [[1, 2, 3], [], [4, 5]] type='3 * var * int64'>
```

In general, if you want to know whether a change in-place is going to change other arrays, such as whether changing this flattened array would change the original, you have to know about the array's layout. Most data analysis should be done at a high-level where you can ignore these differences.

The flattened [ListOffsetArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListOffsetArray.html) has a [NumpyArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.NumpyArray.html) with the same base pointer because flattening a [ListOffsetArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListOffsetArray.html) can be efficiently implemented by just passing or range-slicing its `content`:

```python
>>> a.layout
<ListOffsetArray64>
    <offsets><Index64 i="[0 3 3 5]" offset="0" length="4" at="0x5609caa3af30"/></offsets>
    <content><NumpyArray format="l" shape="5" data="1 2 3 4 5" at="0x5609caa4d050"/></content>
</ListOffsetArray64>
>>> ak.flatten(a).layout
<NumpyArray format="l" shape="5" data="1 2 3 4 5" at="0x5609caa4d050"/>
```

The flattened [ListArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListArray.html) has to, in general, create a new content buffer:

```python
>>> b.layout
<ListArray64>
    <starts><Index64 i="[0 3 3]" offset="0" length="3" at="0x5609ca963990"/></starts>
    <stops><Index64 i="[3 3 5]" offset="0" length="3" at="0x5609ca8f71d0"/></stops>
    <content><NumpyArray format="l" shape="5" data="1 2 3 4 5" at="0x5609caa34590"/></content>
</ListArray64>
>>> ak.flatten(b).layout
<NumpyArray format="l" shape="5" data="1 2 3 4 5" at="0x5609ca460f30"/>
```

Why is that? Because [ListArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListArray.html) `starts` and `stops` don't have to be contiguous and abutting:

```python
>>> c = ak.Array(ak.layout.ListArray64(
...     ak.layout.Index64(np.array([4, 99, 1])),
...     ak.layout.Index64(np.array([7, 99, 3])),
...     ak.layout.NumpyArray(np.array([1000, 4.4, 5.5, 1000, 1.1, 2.2, 3.3, 1000]))
... ))
>>> c
<Array [[1.1, 2.2, 3.3], [], [4.4, 5.5]] type='3 * var * float64'>
```

is a totally legal [ListArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListArray.html), and

```python
>>> ak.flatten(c)
<Array [1.1, 2.2, 3.3, 4.4, 5.5] type='5 * float64'>
>>> ak.flatten(c).layout
<NumpyArray format="d" shape="5" data="1.1 2.2 3.3 4.4 5.5" at="0x5609caa0b710"/>
```

can only make a flattened version of it by cutting pieces out of the original array. (It's not even guaranteed to be _smaller_, since the `starts` and `stops` of [ListArray](https://awkward-array.readthedocs.io/en/latest/ak.layout.ListArray.html) are allowed to _overlap_!)

A future performance optimization might identify ListArrays that are regular enough to be ListOffsetArrays and have [ak.flatten](https://awkward-array.readthedocs.io/en/latest/_auto/ak.flatten.html) return a view. Code that assumes that some operation is a view or a copy can easily be broken by updates in Awkward Array, and that's why I recommend against it.

**Reply by lukasheinrich** [reply]
I tried following this explanation but get `ListOffsetArray` in both cases. Has something changed since the answer?

![image](https://user-images.githubusercontent.com/2318083/115154023-17759180-a079-11eb-9f66-75c1d2718082.png)

**Reply by jpivarski** [reply]
Whether an operation returns a ListArray or a ListOffsetArray is an implementation detail. I can take a closer look later, though.

**Reply by lukasheinrich** [reply]
I was mainly curious/studying up on forms vs types vs layouts and was curious to find any operation that returns a `ListArray` - not high priority

**Reply by jpivarski** [reply]
The types—whether it's a list of not, and whether the list has variable length or a fixed length—is a predictable property of an operation. The form (or for a specific array, the layout) specifies exactly how data are arranged in memory, and that is chosen to optimize performance. It's stable for a given version of Awkward Array, but it's not guaranteed to start the same between versions. (As an example of that, some rearrangements of RecordArrays were made lazy by wrapping them in IndexedArrays.) ListArray and ListOffsetArray differ by whether the list elements of consecutive lists are known to be contiguous in memory. When an operation creates a new output array, it would be a ListOffsetArray, because you might as well make it contiguous, but when a list array is rearranged, it becomes a ListArray because the rearrangement can remain lazy at that level—we can pass through the original `content` with a new `starts` and `stops` to avoid copying (which is why Awkward Arrays aren't mutable in-place).

**Reply by jpivarski** [reply]
I just ran it on a computer and it's true: the operation on `b` has changed so that it's now operating in place. This is the sort of thing that makes it dangerous to _rely_ on the mutability through conversion to NumPy—memory layout in Awkward Array is an implementation detail that can change like this.
